File: formula-cs/Formula/TokenTree/TokenTree.cs
namespace Formula.TokenTree;

public class TokenTree<T>
{
    private readonly RootNode<T> _root = new RootNode<T>();

    public static TokenTree<T> Create() {
        return new TokenTree<T>();
    }

    public TokenTree<T> IgnoreWhitespaces()
    {
        return Add(NodeExpression.AnyOf(CharacterClass.BlankCharacters), 
            _ => default);
    }

    public TokenTree<T> Add(IEnumerable<char> allowed, TokenMapper<T> mapper) {
        return AddBranch(
            NodeExpression.Term(allowed),
            mapper);
    }

    public TokenTree<T> Add(INodeExpression expression, TokenMapper<T> mapper) {
        return AddBranch(
                expression,
                mapper);
    }

    public TokenTree<T> Add(IEnumerable<INodeExpression> expressions, TokenMapper<T> mapper) {
        return AddBranch(
                CompositeNodeExpression.Of(expressions),
                mapper);
    }

    public List<T> Parse(string text) {
        var tokens = new List<T>();
        for (var i = 0; i < text.Length; i++)
        {
            var matches = new List<TokenMatch<T>>();
            _root.Walk(text, i, matches);
            if (matches.Count > 0) 
            {
                var match = matches[0];
                var token = match.Get();
                if (token != null) 
                {
                    tokens.Add(token);
                }
                i = match.EndIndex - 1;
            } 
            else 
            {
                throw new ParseException(
                        GenerateParseErrorMessage(i, text, "did not expect character: '" + text[i] + "'"),
                        text, i);
            }
        }
        return tokens;
    }

    private TokenTree<T> AddBranch(INodeExpression expression, TokenMapper<T> mapper) {
        var node = expression.ChainTo(_root);
        var matcherNode = (MatcherNode<T>) node;
        if (matcherNode.Mapper != null) {
            throw new Exception("Conflicting tokens");
        }

        matcherNode.Mapper = mapper;
        return this;
    }

    private static string GenerateParseErrorMessage(int index, String text, String message) {
        return $"Parse error at index {index} of \"{text}\": {message}" + "\n"
                                                                        + text
                                                                        + "\n"
                                                                        + "^".PadLeft(index);
    }
}

File: formula-java/src/main/java/org/formula/parse/tree/TokenTree.java
package org.formula.parse.tree;

import static org.formula.parse.tree.CharacterClass.BLANK_CHARACTERS;

import java.util.ArrayList;
import java.util.List;

public class TokenTree<T> {
    private final RootNode<T> root = new RootNode<>();

    public static <T> TokenTree<T> create() {
        return new TokenTree<>();
    }

    public TokenTree<T> ignoreWhitespaces() {
        return add(NodeExpression.anyOf(BLANK_CHARACTERS), c -> null);
    }

    public TokenTree<T> add(CharSequence allowed, TokenMapper<T> mapper) {
        return addBranch(
            NodeExpression.term(allowed),
            mapper);
    }

    public TokenTree<T> add(NodeExpression expression, TokenMapper<T> mapper) {
        return addBranch(
                expression,
                mapper);
    }

    public TokenTree<T> add(List<NodeExpression> expressions, TokenMapper<T> mapper) {
        return addBranch(
                CompositeNodeExpression.of(expressions),
                mapper);
    }

    public List<T> parse(String text) {
        List<T> tokens = new ArrayList<>();
        for (int i = 0; i < text.length(); i++) {
            List<TokenMatch<T>> matches = new ArrayList<>();
            root.walk(text, i, matches);
            if (!matches.isEmpty()) {
                var match = matches.get(0);
                var token = match.get();
                if (token != null) {
                    tokens.add(token);
                }
                i = match.endIndex() - 1;
            } else {
                throw new ParseException(
                        generateParseErrorMessage(i, text, "did not expect character: '" + text.charAt(i) + "'"),
                        text, i);
            }
        }
        return tokens;
    }

    private TokenTree<T> addBranch(NodeExpression expression, TokenMapper<T> mapper) {
        Node<T> node = expression.chainTo(root);

        MatcherNode<T> matcherNode = (MatcherNode<T>) node;
        if (matcherNode.mapper() != null) {
            throw new RuntimeException("Conflicting tokens");
        }
        matcherNode.mapper(mapper);
        return this;
    }

    private static String generateParseErrorMessage(int index, String text, String message) {
        return String.format("Parse error at index %d of \"%s\": %s", index, text, message) + "\n"
                + text
                + "\n"
                + " ".repeat(index)
                + "^";
    }
}
